spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/pubsub/GcpPubSubEmulatorConfiguration.java

```diff
@@ -45,7 +45,6 @@
 public class GcpPubSubEmulatorConfiguration {
 
 	@Bean
-	@ConditionalOnMissingBean
```

Comment:
I think you have to reinstate this condition otherwise you don't give end-user any chances to provide some other instance.

Fortunately you did a good fix to the `GcpPubSubAutoConfiguration` already.

Reply:
This `@ConditionalOnMissingBean` isn't behaving as expected.

`GcpPubSubEmulatorConfiguration` is indeed called before `GcpPubSubAutoConfiguration`, but only for the `credentialsProvider` bean. The `transportChannelProvider` bean is obtained from `GcpPubSubAutoConfiguration`.

It's almost as if Spring Boot prioritises beans with less dependencies (`transportChannelProvider` from `GcpPubSubEmulatorConfiguration` takes in `GcpPubSubProperties` whereas `GcpPubSubAutoConfiguration` doesn't take any), and `@ConditionalOnMissingBean` is a second priority criteria...

Reply:
Then it's not right to make your test working breaking the `main` code.
Something is not right indeed. What we have here with the `@ConditionalOnMissingBean` is the way to go. Need more investigation, but I wouldn't merge this change to `master`.

Just IMO, but you are feel free whatever you think is right, even if I think it's not right...

Reply:
It seems like you were right, and the order of the configuration classes in `@SpringBootTest.classes` overrides `@AutoConfigureBefore/After`, which is surprising since `@SpringBootTest`'s Javadoc doesn't mention any config order...